common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+	history struct {
+		lastUpdate         time.Time // only so we know if elapsed times are exceeded, not used to compute per-second rates
+		accepted, rejected PerSecond // requests received, per second (conceptually divided by update rate)
+	}
+
+	// Identity is an arbitrary (stable) identifier for a "limiting" host.
+	Identity string
+	// Limit is the key being used to Limit requests.
+	Limit string
+	// PerSecond represents already-scaled-to-per-second RPS values
+	PerSecond float64
+	// HostWeight is a float between 0 and 1, showing how much of a ratelimit should be allocated to a host.
+	HostWeight float64
+
+	impl struct {
+		cfg Config
+
+		// mut protects usage, as it is the only mutable data
+		mut sync.Mutex
+		// usage data for ratelimits, per host.
+		//
+		// data is first keyed on the limit and then on the host, as it's assumed that there will be
+		// many times more limit keys than limiting hosts, and this will reduce cardinality more rapidly.
+		usage map[Limit]map[Identity]history
+
+		clock clock.TimeSource
+	}
+
+	// Metrics reports overall counts discovered as part of garbage collection.
+	//
+	// This is not necessarily worth maintaining verbatim, but while it's easy to
+	// collect it might give us some insight into overall behavior.
+	Metrics struct {
+		// HostLimits is the number of per-host limits that remain after cleanup
+		HostLimits int
+		// Limits is the number of cross-host limits that remain after cleanup
+		Limits int
+
+		// RemovedHostLimits is the number of per-host limits that were removed as part of this GC pass
+		RemovedHostLimits int
+		// RemovedLimits is the number of cross-host limits that were removed as part of this GC pass
+		RemovedLimits int
+	}
+
+	// Requests contains accepted/rejected request counts for a ratelimit, as part of an update operation.
+	// Only intended to be used in [RequestWeighted.Update].
+	Requests struct {
+		Accepted, Rejected int
+	}
+
+	// RequestWeighted returns aggregated RPS and distribution data for its Limit keys,
+	// based on how many requests each Identity has received.
+	RequestWeighted interface {
+		// Update load-data for this host's requests, given a known elapsed time spent accumulating this load info.
+		//
+		// Elapsed time must be non-zero, but is not otherwise constrained.
+		Update(host Identity, load map[Limit]Requests, elapsed time.Duration)
+
+		// HostWeights returns the per-[Limit] weights for all requested + known keys for this Identity,
+		// as well as the Limit's overall used RPS (to decide RPS to allow for new hosts).
+		HostWeights(host Identity, limits []Limit) (weights map[Limit]HostWeight, usedRPS map[Limit]PerSecond)
+
+		// GC can be called periodically to pre-emptively prune old ratelimits.
+		//
+		// Limit keys that are accessed normally will automatically garbage-collect themselves and old host data,
+		// and an unused limit only costs memory, so this may prove to be unnecessary.
+		GC() Metrics
```

Comment:
Yea, I'm not actually planning on calling this at the moment.  I think the at-read-time pruning is likely to be good enough - memory use isn't huge and there's practically zero CPU cost to completely-unused limits because they're not looked at at all.  And if they're _not_ completely-unused, they'll clean up per-host data automatically, and that'll keep CPU use from growing.

I could remove it, it's more of a pre-emptive "I suspect someone might want this eventually, so it should be possible" -> yea it's easy, might as well just build it.

Reply:
eh, I think I'll add a per-minute call just to emit metrics so we have some "how big is the thing" information.  should be trivial to add to later PRs.